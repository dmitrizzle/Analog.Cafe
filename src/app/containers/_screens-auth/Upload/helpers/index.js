import { ROUTE_SUBMISSION_API } from "../../../../../constants/submission"

// this function kicks user to sign-in scdreen but rembers where to come back to
export const redirectToSignIn = props => {
  props.setLoginRedirectRoutes({ success: props.history.location.pathname })
  props.history.replace({
    pathname: "/sign-in"
  })
}

export const sendSubmission = (data, props) => {
  props.sendUpload({
    method: "post",
    data,
    onUploadProgress: progressEvent => {
      let percentCompleted
      percentCompleted = Math.round(
        progressEvent.loaded * 100 / progressEvent.total
      )
      console.log("Upload percent complete: " + percentCompleted)
    },
    headers: {
      "content-type": "multipart/form-data",
      Authorization: "JWT " + localStorage.getItem("token")
    },
    url: ROUTE_SUBMISSION_API
  })
}
