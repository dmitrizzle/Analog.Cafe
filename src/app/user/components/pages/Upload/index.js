// import "localforage-getitems"

import { connect } from "react-redux"
import {
  loadTextContent,
  loadContent
} from "@roast-cms/french-press-editor/dist/utils/actions-storage"
import React from "react"

import { CARD_ERRORS } from "../../../constants/messages-submission"
import { ROUTE_URL_USER_LANDING } from "../../../constants/routes-session"
import { TEXT_EMOJIS } from "../../../../constants"
import { loadHeader, sendSubmission } from "../../../utils/actions-submission"
import { redirectToSignIn } from "../../../utils/actions-session"
import { resetStatus } from "../../../../admin/store/actions-editor"
import { setModal } from "../../../../core/store/actions-modal"
import { setUserRoutes, resetUserRoutes } from "../../../store/actions-user"
import {
  uploadSubmission,
  resetUploadProgress
} from "../../../store/actions-submission"
import ArticleSection from "../../../../core/components/pages/Article/components/ArticleSection"
import ArticleWrapper from "../../../../core/components/pages/Article/components/ArticleWrapper"
import HeaderSubtitle from "../../../../core/components/vignettes/HeaderLarge/components/HeaderSubtitle"
import HeaderTitle from "../../../../core/components/vignettes/HeaderLarge/components/HeaderTitle"
import HeaderWrapper from "../../../../core/components/vignettes/HeaderLarge/components/HeaderWrapper"
import Link from "../../../../core/components/controls/Link"
import LinkButton from "../../../../core/components/controls/Button/components/LinkButton"
import MetaTags from "../../../../core/components/vignettes/MetaTags"

const STATUS_MESSAGES = {
  pending: "Sending…",
  complete: "Done!",
  error: "Error"
}
class Upload extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      progress: 0,
      status: "pending"
    }
  }
  componentDidMount = () => {
    if (
      !localStorage.getItem("composer-header-state") ||
      JSON.parse(localStorage.getItem("composer-header-state")).title === ""
    ) {
      this.handleEmptySubmission()
      return
    }
    if (
      !localStorage.getItem("composer-content-text") ||
      localStorage.getItem("composer-content-text") === ""
    ) {
      this.handleEmptySubmission()
      return
    }
    if (!localStorage.getItem("token")) {
      redirectToSignIn(this.props)
    } else {
      this.props.resetUserRoutes()
      const submissionConsent = this.props.history.location.pathname.includes(
        "full-consent"
      )
      const content = loadContent()
      const header = loadHeader()
      const textContent = loadTextContent()
      if (!(content && content.document && content.document.nodes)) {
        return this.handleEmptySubmission()
      }
      let data = new FormData()
      data.append("content", JSON.stringify(content))
      data.append("header", JSON.stringify(header))
      data.append("textContent", textContent)
      data.append("isFullConsent", submissionConsent)
      const keys = content.document.nodes
        .filter(node => !!(node.data && node.data.key))
        .map(node => node.data.key)
      const srcs = content.document.nodes
        .filter(node => !!(node.data && node.data.src))
        .map(node => node.data.src)
      if (keys.length > 0) {
        // localForage.getItems(keys).then(results => {
        //   keys.forEach(k => {
        //     data.append("images[" + k + "]", base64ToBlob(results[k]))
        //   })
        //   content.document.nodes
        //     .filter(node => !!(node.data && node.data.src))
        //     .forEach(node => (node.data.src = null))
        //   sendSubmission(data, this.props)
        // })
      } else {
        if (srcs.length === 0) {
          this.props.setModal(
            {
              status: "ok",
              info: CARD_ERRORS.SEND_IMAGES_MISSING
            },
            { url: "errors/submissions" }
          )
          this.props.history.replace({ pathname: "/submit/compose" })
        } else {
          sendSubmission(data, this.props)
        }
      }
    }
  }
  componentWillReceiveProps = nextProps => {
    if (
      nextProps.submission.uploadProgress ===
      this.props.submission.uploadProgress
    )
      return
    if (nextProps.submission.uploadProgress >= 0)
      this.setState({
        progress:
          nextProps.submission.uploadProgress > 0
            ? nextProps.submission.uploadProgress
            : this.state.progress
      })
    else
      this.setState({
        status: "error"
      })
    if (
      this.props.submission.uploadProgress >= 0 &&
      nextProps.submission.uploadProgress === 100
    ) {
      this.props.resetUploadProgress()
      this.props.resetStatus()
      this.setState({
        status: "complete"
      })
    }
  }
  handleEmptySubmission = () => {
    this.props.setModal(
      {
        status: "ok",
        info: CARD_ERRORS.SEND_CONTENT_EMPTY
      },
      { url: "errors/submissions" }
    )
    this.props.history.replace({
      pathname: "/submit/compose"
    })
  }
  render = () => {
    return (
      <ArticleWrapper>
        <MetaTags metaTitle="Sending…" />
        <HeaderWrapper>
          <HeaderTitle>{this.state.progress}%</HeaderTitle>
          <HeaderSubtitle>{STATUS_MESSAGES[this.state.status]}</HeaderSubtitle>
        </HeaderWrapper>
        <ArticleSection>
          {/* <p>
            You have marked your submission as
            <em>
              {this.props.history.location.pathname.includes("full-consent") ===
              true
                ? " open for collaborations. "
                : " closed to collaborations. "}
            </em>
          </p> */}
          {this.state.status === "pending" && (
            <p>
              <strong>
                <span role="img" aria-label="Warning">
                  {TEXT_EMOJIS.WARNING}
                </span>{" "}
                Please keep this page open, do not refresh and do not click your
                browser’s “back” button
              </strong>{" "}
              while your submission is sending (uploading).
            </p>
          )}
          {this.state.status === "error" && (
            <p>
              <strong>{CARD_ERRORS.SEND.text}</strong>{" "}
              <a
                href="#reload"
                onClick={event => {
                  event.preventDefault()
                  window.location.reload()
                }}
              >
                Try again
              </a>{" "}
              or <Link to={"/submit/compose"}>go back</Link>.
            </p>
          )}
          {this.state.status === "complete" && (
            <div>
              <MetaTags metaTitle="Done!" />
              <p>
                We’ve received your work. It’ll take a couple of minutes to
                process the images – after that, you should be able to see it{" "}
                <Link to={ROUTE_URL_USER_LANDING}>here</Link>.
              </p>
              <p>Thank you so much for your contribution!</p>
              <LinkButton branded to={ROUTE_URL_USER_LANDING}>
                My Submissions
              </LinkButton>
            </div>
          )}
        </ArticleSection>
      </ArticleWrapper>
    )
  }
}
const mapStateToProps = state => {
  return {
    composer: state.composer,
    submission: state.submission,
    editor: state.editor,
    user: state.user
  }
}
const mapDispatchToProps = dispatch => {
  return {
    uploadSubmission: request => {
      dispatch(uploadSubmission(request))
    },
    resetUploadProgress: () => {
      dispatch(resetUploadProgress())
    },
    resetStatus: () => {
      dispatch(resetStatus())
    },
    setUserRoutes: routes => {
      dispatch(setUserRoutes(routes))
    },
    resetUserRoutes: () => {
      dispatch(resetUserRoutes())
    },
    setModal: (info, request) => {
      dispatch(setModal(info, request))
    }
  }
}
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Upload)
