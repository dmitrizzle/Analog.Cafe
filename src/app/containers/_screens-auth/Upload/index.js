// tools
import React from "react"
import Helmet from "react-helmet"

import { connect } from "react-redux"

import localForage from "localforage"
import "localforage-getitems"

// components
import Heading from "../../../components/ArticleHeading"
import { Article, Section } from "../../../components/ArticleStyles"

// constants & helpers
import { loadContent, loadHeader } from "../../Composer/helpers/loader"
import { WEBSOCKET_UPLOAD_PROGRESS } from "../../../../constants/submission"
import { ROUTE_REDIRECT_AFTER_SUBMIT } from "../../../../constants/submission"
import emojis from "../../../../constants/messages/emojis"

import { redirectToSignIn, sendSubmission } from "./helpers"

// redux actions
import {
  setRoutes as setLoginRedirectRoutes,
  resetRoutes as resetLoginRedirectRoutes
} from "../../../../actions/userActions"
import {
  send as sendUpload,
  initStatus as resetUploadStatus
} from "../../../../actions/uploadActions"

// render
class Upload extends React.PureComponent {
  constructor(props) {
    super(props)
    this.socketUpload = new WebSocket(WEBSOCKET_UPLOAD_PROGRESS)
  }
  componentDidMount = () => {
    // redirects
    if (!localStorage.getItem("token")) {
      redirectToSignIn(this.props)
    } else {
      this.props.resetLoginRedirectRoutes()

      // WebSocket
      this.socketUpload.addEventListener("message", function(event) {
        console.log("Upload image form server to cloundinary: " + event.data)
      })

      // construct submission data
      const submissionConsent = this.props.history.location.pathname.includes(
        "full-consent"
      )
      const content = loadContent()
      const header = loadHeader()
      let keys = []
      if (content && content.document && content.document.nodes) {
        // form adata obj
        let data = new FormData()
        data.append("content", JSON.stringify(content))
        data.append("header", JSON.stringify(header))
        data.append("isFullConsent", submissionConsent)

        // get filename keys for the saved images in the submission
        for (var i = 0; i < content.document.nodes.length; i++) {
          var node = content.document.nodes[i]
          if (node && node.data && node.data.key) {
            keys.push(node.data.key)
          }
        }
        if (keys.length > 0) {
          const _this = this // perhaps binding "this" to localForage fn would have been the right decision instead of _this - but I don't know and don't care to do this atm
          localForage.getItems(keys).then(function(results) {
            for (var i = 0; i < keys.length; i++) {
              data.append("images[" + keys[i] + "]", results[keys[i]])
            }
            sendSubmission(data, _this.props)
          })
        } else {
          console.log("no images")
          sendSubmission(data, this.props)
        }
      } else {
        // content deleted
        // TODO notify users that content is deleted or missplaced
        this.props.history.replace({
          pathname: ROUTE_REDIRECT_AFTER_SUBMIT
        })
      }
    }
  }

  componentWillReceiveProps = () => {
    if (this.props.upload.status === "ok") {
      // clear submissions content and image in storage
      localStorage.removeItem("composer-content-state")
      localStorage.removeItem("composer-header-state")
      localForage.clear()
      // close connection listener
      this.socketUpload.close()
      // reset upload state
      this.props.resetUploadStatus()
      // redirect after submission complete
      this.props.history.replace({
        pathname: ROUTE_REDIRECT_AFTER_SUBMIT
      })
    } else if (this.props.upload.status !== "pending") {
      // submission not in progress
      this.socketUpload.close() // close socket connection
      // if user is unauthorized, redirect to sign in page
      if (this.props.upload.status === "unauthorized") {
        redirectToSignIn(this.props)
      }
    }
  }

  render = () => {
    return (
      <Article>
        <Helmet>
          <title>Uploading Submissoin…</title>
        </Helmet>
        <Heading pageTitle={emojis.NEONCAT} pageSubtitle="Sending…" />
        <Section>
          <p>
            You have requested
            <em>
              {this.props.history.location.pathname.includes("full-consent") ===
              true ? (
                " more exposure. "
              ) : (
                " no additional exposure. "
              )}
            </em>
            Please wait while we process your submission&hellip;
          </p>
        </Section>
      </Article>
    )
  }
}

// connet with redux
const mapStateToProps = state => {
  return {
    user: state.user,
    upload: state.upload
  }
}
const mapDispatchToProps = dispatch => {
  return {
    sendUpload: request => {
      dispatch(sendUpload(request))
    },
    resetUploadStatus: () => {
      dispatch(resetUploadStatus())
    },
    setLoginRedirectRoutes: routes => {
      dispatch(setLoginRedirectRoutes(routes))
    },
    resetLoginRedirectRoutes: () => {
      dispatch(resetLoginRedirectRoutes())
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Upload)
// NOTE: withRouter() props inherited from /containers/_screens/AppRoutesSubmit
