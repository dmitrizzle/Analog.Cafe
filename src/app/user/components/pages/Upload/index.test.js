import { createMockStore } from "redux-test-utils"
import { shallowWithStore } from "enzyme-redux"
import React from "react"

import Upload from "./"

it("Render Upload without crashing", () => {
  const store = createMockStore()
  shallowWithStore(<Upload />, store)
})
