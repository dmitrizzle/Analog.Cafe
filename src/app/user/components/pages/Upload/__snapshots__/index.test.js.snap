// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Upload, matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Upload) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "composer": undefined,
      "editor": undefined,
      "resetStatus": [Function],
      "resetUploadProgress": [Function],
      "resetUserRoutes": [Function],
      "setModal": [Function],
      "setUserRoutes": [Function],
      "submission": undefined,
      "uploadSubmission": [Function],
      "user": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "composer": undefined,
        "editor": undefined,
        "resetStatus": [Function],
        "resetUploadProgress": [Function],
        "resetUserRoutes": [Function],
        "setModal": [Function],
        "setUserRoutes": [Function],
        "submission": undefined,
        "uploadSubmission": [Function],
        "user": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
